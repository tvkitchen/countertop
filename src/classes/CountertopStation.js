import { v4 as uuid } from 'uuid'
import { countertopStates } from '../constants'
import { consoleLogger } from '../tools/loggers'
import { CountertopWorker } from './CountertopWorker'

class CountertopStation {
	logger = null

	kafka = null

	id = null

	Appliance = null

	applianceSettings = null

	workers = []

	state = ''

	/**
	 * Create a CountertopStation
	 *
	 * @param  {Class}  Appliance         The IAppliance class that this station will manage.
	 * @param  {Object} applianceSettings Settings for the appliance.
	 * @param  {Logger} options.logger    A logger with methods for all TV Kitchen logLevels.
	 */
	constructor(
		Appliance,
		applianceSettings = {},
		{
			logger = consoleLogger,
			kafka,
		} = {},
	) {
		if (Appliance === undefined) {
			throw new Error('CountertopStation requires an Appliance.')
		}
		this.Appliance = Appliance
		this.applianceSettings = applianceSettings
		this.logger = logger
		this.kafka = kafka
		this.id = `${Appliance.name}::${uuid()}`
		this.setState(countertopStates.STOPPED)
	}

	/**
	 * Replace the station's workers in order to fit a specified topology.
	 *
	 * @param  {CountertopToplogy} topology The topology to use.
	 * @return {CountertopWorker[]}         The new set of workers.
	 */
	invokeTopology = (topology) => {
		this.logger.trace(`CountertopStation<${this.id}>: invokeTopology()`)
		if (this.getState() !== countertopStates.STOPPED) {
			throw new Error('CountertopStations must be stopped before invoking a topology.')
		}
		const stationStreams = topology.streams.filter((stream) => stream.getMouth() === this)
		this.workers = stationStreams.map((stream) => new CountertopWorker(
			this.Appliance,
			this.applianceSettings,
			{
				stream,
				logger: this.logger,
				kafka: this.kafka,
			},
		))
		return this.workers
	}

	/**
	 * Start the station and begin data processing.
	 *
	 * @return {Boolean} Whether the station successfully started.
	 */
	start = async () => {
		this.logger.trace(`CountertopStation<${this.id}>: start()`)
		this.state = countertopStates.STARTING
		const promises = this.workers.map(async (worker) => worker.start())
		const started = (await Promise.all(promises)).every((result) => result)
		this.setState(started ? countertopStates.STARTED : countertopStates.ERRORED)
		return this.getState() === countertopStates.STARTED
	}

	/**
	 * Stop the countertop and halt all data processing.
	 *
	 * @return {Boolean} Whether the station successfully stopped.
	 */
	stop = async () => {
		this.logger.trace(`CountertopStation<${this.id}>: stop()`)
		this.setState(countertopStates.STOPPING)
		const promises = this.workers.map(async (worker) => worker.stop())
		const stopped = (await Promise.all(promises)).every((result) => result)
		this.setState(stopped ? countertopStates.STOPPED : countertopStates.ERRORED)
		return this.getState() === countertopStates.STOPPED
	}

	/**
	 * Get the data types that this station processes.
	 *
	 * This is determined by the Appliance that the station hosts.
	 *
	 * @return {String[]} The data types that this station processes.
	 */
	getInputTypes = () => this.Appliance.getInputTypes(this.applianceSettings)

	/**
	 * Get the data types that this station produces.
	 *
	 * This is determined by the Appliance that the station hosts.
	 *
	 * @return {String[]} The data types that this station produces.
	 */
	getOutputTypes = () => this.Appliance.getOutputTypes(this.applianceSettings)

	/**
	 * Get the current state of the CountertopStation.
	 *
	 * @return {String} The state of the CountertopStation.
	 */
	getState = () => this.state

	/**
	 * Set the current state of the CountertopStation.
	 *
	 * This is an internal method.
	 */
	setState = (state) => { this.state = state }

	/**
	 * Registers a listener to the CountertopStation for a given event type.
	 *
	 * Event types are defined in @tvkitchen/base-constants
	 *
	 * @param  {String} eventType  The type of event being listened to.
	 * @param  {Function} listener The listener to be registered for that event.
	 * @return {CountertopStation} The CountertopStation instance (to enable chaining).
	 */
	on = (eventType, listener) => {
		this.workers.forEach((worker) => worker.on(eventType, listener))
		return this
	}
}

export default CountertopStation
