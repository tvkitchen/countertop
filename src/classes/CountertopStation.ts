import { v4 as uuid } from 'uuid'
import { Kafka } from 'kafkajs'
import { CountertopStationState } from '../types'
import { CountertopStationStateError } from '../errors'
import { CountertopWorker } from './CountertopWorker'
import type { Logger } from '../types'
import type {
	ImplementedApplianceClass,
	ApplianceSettings,
} from './AbstractAppliance'
import type { CountertopTopology } from './CountertopTopology'

interface CountertopStationSettings {
	logger?: Logger;
	kafka: Kafka;
}

export class CountertopStation {
	public readonly settings: CountertopStationSettings

	public readonly id: string

	public readonly ApplianceClass: ImplementedApplianceClass

	public readonly applianceSettings?: ApplianceSettings

	#workers: CountertopWorker[]

	#state: CountertopStationState

	constructor(
		ApplianceClass: ImplementedApplianceClass,
		applianceSettings?: ApplianceSettings,
		settings: CountertopStationSettings = {
			kafka: new Kafka({ brokers: [] }),
		},
	) {
		this.ApplianceClass = ApplianceClass
		this.applianceSettings = applianceSettings
		this.settings = settings
		this.id = `${ApplianceClass.name}::${uuid()}`
		this.#state = CountertopStationState.Stopped
		this.#workers = []
	}

	/**
	 * Replace the station's workers in order to fit a specified topology.
	 *
	 * @param  {CountertopToplogy} topology The topology to use.
	 * @return {CountertopWorker[]}         The new set of workers.
	 */
	public invokeTopology(
		topology: CountertopTopology,
	): CountertopWorker[] {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: invokeTopology()`)
		if (this.state !== CountertopStationState.Stopped) {
			throw new CountertopStationStateError('CountertopStations must be stopped when invoking a topology.')
		}

		const stationStreams = topology.streams.filter(
			(stream) => stream.mouth === this,
		)
		this.#workers = stationStreams.map((stream) => (
			new CountertopWorker(
				this.ApplianceClass,
				this.applianceSettings,
				{
					stream,
					logger: this.settings.logger,
					kafka: this.settings.kafka,
				},
			)
		))
		return this.#workers
	}

	/**
	 * Start the CountertopStation and begin data processing.
	 */
	public async start(): Promise<void> {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: start()`)
		this.#state = CountertopStationState.Starting
		try {
			await this.startAllWorkers()
		} catch (e: unknown) {
			this.#state = CountertopStationState.Errored
			await this.stopAllWorkers()
			throw e
		}
		this.#state = CountertopStationState.Started
	}

	/**
	 * Stop the CountertopStation and halt all data processing.
	 */
	public async stop(): Promise<void> {
		this.settings.logger?.debug(`CountertopStation<${this.id}>: stop()`)
		this.#state = CountertopStationState.Stopping
		await this.stopAllWorkers()
	}

	/**
	 * Get the data types that this station processes.
	 *
	 * This is determined by the Appliance that the station hosts.
	 */
	public getInputTypes() {
		return this.ApplianceClass.getInputTypes(this.applianceSettings)
	}

	/**
	 * Get the data types that this station produces.
	 *
	 * This is determined by the Appliance that the station hosts.
	 */
	public getOutputTypes() {
		return this.ApplianceClass.getOutputTypes(this.applianceSettings)
	}

	public get state() {
		return this.#state
	}

	/**
	 * Registers a listener to the CountertopStation for a given event type.
	 *
	 * Event types are defined in @tvkitchen/base-constants
	 */
	public on(
		eventType: string,
		listener: (...args: unknown[]) => void,
	): this {
		this.#workers.forEach(
			(worker) => worker.on(eventType, listener),
		)
		return this
	}

	private async startAllWorkers(): Promise<void> {
		await Promise.all(
			this.#workers.map(
				async (worker) => worker.start(),
			),
		)
	}

	private async stopAllWorkers(): Promise<void> {
		await Promise.all(
			this.#workers.map(
				async (worker) => worker.stop(),
			),
		)
	}
}
