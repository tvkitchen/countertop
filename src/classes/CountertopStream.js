import assert from 'assert'
import { v4 as uuid } from 'uuid'
import {
	getLongestStreamLength,
	getSourcesFromStreamMap,
} from '../tools/utils/countertop'

class CountertopStream {
	id = null

	tributaries = null

	mouth = null

	source = null

	/**
	 * Create a new CountertopStream.
	 *
	 * CountertopStreams are made up of a distinct stream of stations (e.g. Appliance instances).
	 * This allows a given countertop to have multiple sources of video data being processed
	 * simultaneously.
	 *
	 * The tributary map does not need to be complete -- there can be inputs with no associated
	 * tributary streams.
	 *
	 * The tributary map cannot contain irrelevant streams (the output map should match the inputs).
	 *
	 * @param  {CountertopStation} station     The station at the end of the stream.
	 * @param  {Map}               tributaries The streams that are flowing into the final station.
	 */
	constructor(station, tributaries = new Map()) {
		Array.from(tributaries.keys()).forEach(
			(outputType) => assert(
				station.getInputTypes().includes(outputType),
				'The tributaries contained an invalid stream type.',
			),
		)
		this.id = `CountertopStream::${uuid()}`
		this.mouth = station
		this.tributaries = tributaries
		if (station.getInputTypes().length === 0) {
			this.source = station
		} else {
			this.source = getSourcesFromStreamMap(tributaries).pop() ?? null
		}
	}

	/**
	 * Get the origin / first station in the stream.
	 *
	 * @return {CountertopStation} The origin station in the stream.
	 */
	getSource = () => this.source

	/**
	 * Get the last station in the stream.
	 * This is the station that will produce new data, and determines the stream's
	 * output types.
	 *
	 * @return {CountertopStation} The last station in the stream.
	 */
	getMouth = () => this.mouth

	/**
	 * Get the Map of tributary streams that feed into the mouth station.
	 *
	 * @return {Map} The map of streams keyed by the input types they feed into.
	 */
	getTributaryMap = () => this.tributaries

	/**
	 * Get an array of tributaries that feed into the mouth station.
	 *
	 * @return {CountertopStream[]} The array of tributary streams.
	 */
	getTributaryArray = () => [...this.tributaries.values()]

	/**
	 * Get the maximum length of the longest stream path between the stream's source and mouth.
	 *
	 * @return {Number} The length of the stream
	 */
	getLength = () => getLongestStreamLength(this.getTributaryArray()) + 1

	/**
	 * Checks if a given station exists at any point within the stream.
	 *
	 * @param  {CountertopStation} station The station being searched for.
	 * @return {Boolean}                   The result of the search.
	 */
	includesStation = (station) => this.mouth === station
		|| this.getTributaryArray().some((tributary) => tributary.includesStation(station))

	/**
	 * Get the output types produced by this stream.
	 *
	 * @return {String[]} The output types.
	 */
	getOutputTypes = () => this.mouth.getOutputTypes()
}

export default CountertopStream
