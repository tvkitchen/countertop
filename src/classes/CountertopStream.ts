import assert from 'assert'
import { v4 as uuid } from 'uuid'
import {
	getLongestStreamLength,
	getSourcesFromStreams,
} from '../tools/utils/countertop'

export class CountertopStream {
	public readonly id: string

	public readonly tributaries: Map<string, CountertopStream>

	public readonly mouth: any // this will be a CountertopStation once defined

	public readonly source: any // this will be a CountertopStation once defined

	/**
	 * Create a new CountertopStream.
	 *
	 * CountertopStreams are made up of a distinct stream of stations (e.g. Appliance instances).
	 * This allows a given Countertop to have multiple sources of video data being processed
	 * simultaneously.
	 *
	 * The tributary map does not need to be complete -- there can be inputs with no associated
	 * tributary streams.
	 *
	 * The tributary map cannot contain irrelevant streams (the output map should match the inputs).
	 *
	 * @param  {CountertopStation} station The station at the end of the stream.
	 * @param  {Map<string, CountertopStream>} tributaries The streams providing data for each input.
	 */
	constructor(
		countertopStation: any, // this will be a CountertopStation once defined
		tributaries = new Map<string, CountertopStream>(),
	) {
		Array.from(tributaries.keys()).forEach(
			(outputType) => assert(
				// TODO: remove this when CountertopStation has type definitions
				// eslint-disable-next-line
				countertopStation.getInputTypes().includes(outputType),
				'The tributaries contained an invalid stream type.',
			),
		)
		this.id = `CountertopStream::${uuid()}`
		// TODO: remove this when CountertopStation has type definitions
		// eslint-disable-next-line
		this.mouth = countertopStation
		this.tributaries = tributaries
		// TODO: remove this when CountertopStation has type definitions
		// eslint-disable-next-line
		if (countertopStation.getInputTypes().length === 0) {
			// TODO: remove this when CountertopStation has type definitions
			// eslint-disable-next-line
			this.source = countertopStation
		} else {
			// TODO: remove this when CountertopStation has type definitions
			// eslint-disable-next-line
			this.source = getSourcesFromStreams(this.getTributaryArray()).pop() ?? null
		}
	}

	/**
	 * Get an array of tributaries that feed into the mouth station.
	 *
	 * @return {CountertopStream[]} The array of tributary streams.
	 */
	public getTributaryArray(): CountertopStream[] {
		return [...this.tributaries.values()]
	}

	/**
	 * Get the maximum length of the longest stream path between the stream's source and mouth.
	 *
	 * @return {Number} The length of the stream
	 */
	public getLength() {
		return getLongestStreamLength(this.getTributaryArray()) + 1
	}

	/**
	 * Checks if a given station exists at any point within the stream.
	 *
	 * @param  {CountertopStation} station The station being searched for.
	 * @return {Boolean}                   The result of the search.
	 */
	public includesStation(station: any): boolean {
		return this.mouth === station
			|| this.getTributaryArray().some(
				// TODO: remove this when CountertopStation has type definitions
				// eslint-disable-next-line
				(tributary: any) => tributary.includesStation(station),
			)
	}

	/**
	 * Checks if a given sub-stream exists at any point within the stream.
	 */
	public includesStream(subStream: CountertopStream): boolean {
		return this.getTributaryArray().includes(subStream)
			|| this.getTributaryArray().some(
				(tributary: CountertopStream): boolean => tributary.includesStream(subStream),
			)
	}

	/**
	 * Get the output types produced by this stream.
	 *
	 * @return {String[]} The output types.
	 */
	public getOutputTypes(): any {
		// TODO: remove this when CountertopStation has type definitions
		// eslint-disable-next-line
		return this.mouth.getOutputTypes()
	}
}
