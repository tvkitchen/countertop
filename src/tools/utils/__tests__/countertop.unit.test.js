import { by } from '..'
import CountertopStation from '../../../classes/CountertopStation'
import CountertopStream from '../../../classes/CountertopStream'
import {
	generateMockAppliance,
	normalizeTributaryMaps,
} from '../jest'
import {
	getSourceStations,
	getLongestStreamLength,
	getCollectiveOutputTypes,
	getStationsThatConsumeTypes,
	getStreamsThatProduceTypes,
	filterStreamsContainingStation,
	getStreamOutputMap,
	getSourcesFromStreams,
	getStreamsFromStreamMap,
	getSourcesFromStreamMap,
	generateTributaryMaps,
} from '../countertop'

describe('countertop', () => {
	describe('getSourceStations', () => {
		it('Should extract stations that have no inputs', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			}))
			const stations = [
				stationA,
				stationB,
			]
			const sourceStations = getSourceStations(stations)
			expect(sourceStations).toEqual([stationA])
		})
	})

	describe('getLongestStreamLength', () => {
		it('should return the length of the longest stream', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streams = [
				streamA,
				streamB,
			]
			expect(getLongestStreamLength(streams)).toBe(2)
		})
	})

	describe('getCollectiveOutputTypes', () => {
		it('should return the output types of multiple streams', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streams = [
				streamA,
				streamB,
			]
			const collectiveOutputTypes = getCollectiveOutputTypes(streams)
			expect(collectiveOutputTypes.sort()).toEqual(
				['foo', 'bar', 'baz'].sort(),
			)
		})
	})

	describe('getStationsThatConsumeTypes', () => {
		it('should return the stations that consume the specified types', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: ['red'],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const stations = [
				stationA,
				stationB,
			]
			const result = getStationsThatConsumeTypes(stations, ['foo'])
			expect(result).toEqual([stationB])
		})
	})

	describe('getStreamsThatProduceTypes', () => {
		it('should return the streams that produce the specified types', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streamC = new CountertopStream(stationA)
			const streams = [
				streamA,
				streamB,
				streamC,
			]
			const result = getStreamsThatProduceTypes(streams, ['foo'])
			expect(result.sort(by('id'))).toEqual([streamA, streamC].sort(by('id')))
		})
	})

	describe('filterStreamsContainingStation', () => {
		it('should remove streams that contain a given station', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const stationC = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['bloo'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streamC = new CountertopStream(stationC)
			const streams = [
				streamA,
				streamB,
				streamC,
			]
			const result = filterStreamsContainingStation(streams, stationA)
			expect(result).toEqual([streamC])
		})
	})

	describe('getStreamOutputMap', () => {
		it('should map streams by their output types', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streams = [
				streamA,
				streamB,
			]
			const result = getStreamOutputMap(streams)
			expect(result).toEqual(new Map([
				['foo', [streamA]],
				['bar', [streamB]],
				['baz', [streamB]],
			]))
		})
	})

	describe('getSourcesFromStreams', () => {
		it('should return all sources represented by streams', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const stationC = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streamC = new CountertopStream(stationC)
			const streams = [
				streamB,
				streamC,
			]
			const result = getSourcesFromStreams(streams)
			expect(result.sort(by('id'))).toEqual([stationA, stationC].sort(by('id')))
		})
	})

	describe('getStreamsFromStreamMap', () => {
		it('should return the streams represented in a stream map', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const stationC = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streamC = new CountertopStream(stationC)
			const streamMap = new Map([
				['foo', [streamA, streamC]],
				['bar', [streamB]],
				['baz', [streamB]],
			])
			const result = getStreamsFromStreamMap(streamMap)
			expect(result.sort(by('id'))).toEqual(
				[streamA, streamB, streamC].sort(by('id')),
			)
		})
	})

	describe('getSourcesFromStreamMap', () => {
		it('should return the set of sources represented in a stream map', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const stationC = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streamC = new CountertopStream(stationC)
			const streamMap = new Map([
				['foo', [streamA, streamC]],
				['bar', [streamB]],
				['baz', [streamB]],
			])
			const result = getSourcesFromStreamMap(streamMap)
			expect(result.sort(by('id'))).toEqual(
				[stationA, stationC].sort(by('id')),
			)
		})
	})

	describe('generateTributaryMaps', () => {
		it('should not create tributary maps of streams with different sources', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo'],
				outputTypes: ['bar', 'baz'],
			}))
			const stationC = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationD = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo', 'bar'],
				outputTypes: ['baz'],
			}))
			const stations = [
				stationA,
				stationB,
				stationC,
				stationD,
			]
			const streamA = new CountertopStream(stationA)
			const streamB = new CountertopStream(stationB, new Map([['foo', streamA]]))
			const streamC = new CountertopStream(stationC)
			const streamMap = new Map([
				['foo', [streamA, streamC]],
				['bar', [streamB]],
				['baz', [streamB]],
			])
			const result = generateTributaryMaps(stationD, streamMap)
			expect(normalizeTributaryMaps(result, stations)).toMatchSnapshot()
		})

		it('should return a partially populated tributary map if there are not enough outputs for all inputs', () => {
			const stationA = new CountertopStation(generateMockAppliance({
				inputTypes: [],
				outputTypes: ['foo'],
			}))
			const stationB = new CountertopStation(generateMockAppliance({
				inputTypes: ['foo', 'bar'],
				outputTypes: ['baz'],
			}))
			const stations = [
				stationA,
				stationB,
			]
			const streamA = new CountertopStream(stationA)
			const streamMap = new Map([
				['foo', [streamA]],
			])
			const result = generateTributaryMaps(stationB, streamMap)
			expect(normalizeTributaryMaps(result, stations)).toMatchSnapshot()
		})
	})
})
