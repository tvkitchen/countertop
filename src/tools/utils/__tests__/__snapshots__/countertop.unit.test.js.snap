// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`countertop generateTributaryMaps should not create tributary maps of streams with different sources 1`] = `
Array [
  Map {
    "foo" => Object {
      "mouth": 0,
      "source": 0,
      "tributaryMap": Map {},
    },
    "bar" => Object {
      "mouth": 1,
      "source": 0,
      "tributaryMap": Map {
        "foo" => Object {
          "mouth": 0,
          "source": 0,
          "tributaryMap": Map {},
        },
      },
    },
  },
  Map {
    "foo" => Object {
      "mouth": 2,
      "source": 2,
      "tributaryMap": Map {},
    },
  },
]
`;

exports[`countertop generateTributaryMaps should return a partially populated tributary map if there are not enough outputs for all inputs 1`] = `
Array [
  Map {
    "foo" => Object {
      "mouth": 0,
      "source": 0,
      "tributaryMap": Map {},
    },
  },
]
`;
